components/brave_wallet/browser/nft_metadata_fetcher.cc

```diff
@@ -0,0 +1,376 @@
+                               std::move(internal_callback));
+}
+
+void NftMetadataFetcher::OnSanitizeTokenMetadata(
+    GetTokenMetadataIntermediateCallback callback,
+    data_decoder::JsonSanitizer::Result result) {
+  if (result.error) {
+    VLOG(1) << "Data URI JSON validation error:" << *result.error;
+    std::move(callback).Run(
+        "", static_cast<int>(mojom::JsonRpcError::kParsingError),
+        l10n_util::GetStringUTF8(IDS_WALLET_PARSING_ERROR));
+    return;
+  }
+
+  std::string metadata_json;
+  if (result.value.has_value()) {
+    metadata_json = result.value.value();
+  }
+
+  std::move(callback).Run(metadata_json, 0, "");  // 0 is kSuccess
+}
+
+void NftMetadataFetcher::OnGetTokenMetadataPayload(
+    GetTokenMetadataIntermediateCallback callback,
+    APIRequestResult api_request_result) {
+  mojom::ProviderErrorUnionPtr error;
+  if (!api_request_result.Is2XXResponseCode()) {
+    std::move(callback).Run(
+        "", static_cast<int>(mojom::JsonRpcError::kInternalError),
+        l10n_util::GetStringUTF8(IDS_WALLET_INTERNAL_ERROR));
+    return;
+  }
+
+  // Invalid JSON becomes an empty string after sanitization
+  if (api_request_result.body().empty()) {
+    std::move(callback).Run(
+        "", static_cast<int>(mojom::JsonRpcError::kParsingError),
+        l10n_util::GetStringUTF8(IDS_WALLET_PARSING_ERROR));
+    return;
+  }
+
+  std::move(callback).Run(api_request_result.body(), 0, "");  // 0 is kSuccess
+}
+
+void NftMetadataFetcher::CompleteGetEthTokenMetadata(
+    GetEthTokenMetadataCallback callback,
+    const std::string& response,
+    int error,
+    const std::string& error_message) {
+  std::move(callback).Run(response, mojom::ProviderError(error), error_message);
+}
+
+void NftMetadataFetcher::GetSolTokenMetadata(
+    const std::string& nft_account_address,
+    GetSolTokenMetadataCallback callback) {
+  // Derive metadata PDA for the NFT accounts
+  absl::optional<std::string> associated_metadata_account =
+      SolanaKeyring::GetAssociatedMetadataAccount(nft_account_address);
+  if (!associated_metadata_account) {
+    std::move(callback).Run(
+        "", mojom::SolanaProviderError::kInternalError,
+        l10n_util::GetStringUTF8(IDS_WALLET_INTERNAL_ERROR));
+    return;
+  }
+
+  auto internal_callback =
+      base::BindOnce(&NftMetadataFetcher::OnGetSolanaAccountInfoTokenMetadata,
+                     weak_ptr_factory_.GetWeakPtr(), std::move(callback));
+  json_rpc_service_->GetSolanaAccountInfo(*associated_metadata_account,
+                                          std::move(internal_callback));
+}
+
+void NftMetadataFetcher::OnGetSolanaAccountInfoTokenMetadata(
+    GetSolTokenMetadataCallback callback,
+    absl::optional<SolanaAccountInfo> account_info,
+    mojom::SolanaProviderError error,
+    const std::string& error_message) {
+  if (error != mojom::SolanaProviderError::kSuccess || !account_info) {
+    std::move(callback).Run("", std::move(error), error_message);
```

Comment:
Thanks. My understanding is that std::move() is not necessary in this case because mojom types do not define copy constructors, so a move constructor here is used by default because the parameter is a value type.

Reply:
It's just `enum class SolanaProviderError : int32_t` in the generated `brave_wallet.mojom-shared.h`.